system/include/simd128.h

```diff
@@ -0,0 +1,747 @@
+/*
+WebAssembly SIMD128 Intrinsics
+*/
+
+#include <stdint.h>
+
+// User-facing types
+typedef int8_t v128 __attribute__((__vector_size__(16)));
+typedef int8_t i8x16 __attribute__((__vector_size__(16)));
+typedef uint8_t u8x16 __attribute__((__vector_size__(16)));
+typedef int16_t i16x8 __attribute__((__vector_size__(16)));
+typedef uint16_t u16x8 __attribute__((__vector_size__(16)));
+typedef int32_t i32x4 __attribute__((__vector_size__(16)));
+typedef uint32_t u32x4 __attribute__((__vector_size__(16)));
+typedef float f32x4 __attribute__((__vector_size__(16)));
+
+#ifdef __wasm_unimplemented_simd128__
+
+typedef int64_t i64x2 __attribute__((__vector_size__(16)));
+typedef uint64_t u64x2 __attribute__((__vector_size__(16)));
+typedef double f64x2 __attribute__((__vector_size__(16)));
+
+#endif // __wasm_unimplemented_simd128__
+
+// Internal types
+typedef char __v128 __attribute__((__vector_size__(16)));
+typedef char  __i8x16 __attribute__((__vector_size__(16)));
+typedef short __i16x8 __attribute__((__vector_size__(16)));
+typedef int __i32x4 __attribute__((__vector_size__(16)));
+typedef long long __i64x2 __attribute__((__vector_size__(16)));
+
+
+#define __DEFAULT_FN_ATTRS __attribute__((__always_inline__, __nodebug__))
+
+// v128 wasm_v128_load(v128* mem)
+static __inline__ v128 __DEFAULT_FN_ATTRS wasm_v128_load(v128* mem) {
+  return (*mem);
+}
+
+// wasm_v128_store(v128 *mem, v128 a)
+static __inline__ void __DEFAULT_FN_ATTRS wasm_v128_store(v128* mem, v128 a) {
+  *mem = a;
+}
```

Comment:
How do unaligned load and stores work ?

Reply:
Also, if someone does, e.g.,

```c++
float* ptr = ...not aligned to a v128 boundary...;
v128 val = wasm_v128_load((v128*)ptr); // UB
```

the behavior is undefined, because `wasm_v128_load` will perform an aligned load on an unaligned address. This feels like a footgun, given that the WASM instruction does support unaligned loads just fine. 

May be worth it to call this out, at least in a comment, and maybe also, e.g., by using `_aligned`/`_unaligned` suffixes to make that clearer.

Reply:
Currently they do not. The underlying WebAssembly architecture does support unaligned loads and stores, of course, but using them may have silent and very large performance problems on some platforms, so we definitely don't want to encourage their use. Given that, I wonder how important it is to allow for unaligned loads and stores in this API.

Reply:
I think it is reasonable to match the builtin, and document this somehow, unless you are going to change the builtins before merging this in.

By the way, what happens with the alignment hint in practice? Do some implementations do something different for different alignment (aside from checking it)?

Reply:
@penzn Are implementations required to check the alignment?

Reply:
The expectation is that on platforms that trap on unaligned accesses, engines would be allowed to catch those traps and fix up the operation afterward. In other words, every unaligned access could come with the cost of a trap through the OS. I do not know how engines use the alignment hints, but at the tools level we want to minimize unaligned accesses because the alignment hints are only hints.

Reply:
@gnzlbg Yes, alignment less than natural is a runtime error.

Reply:
> The expectation is that on platforms that trap on unaligned accesses, engines would be allowed to catch those traps and fix up the operation afterward.

Which platforms don't have unaligned vector load instructions?

Reply:
@penzn I believe unaligned accesses are technically allowed in the spec proposal, just as they are for all other value types. We just don't want to encourage using them in the tools.

@gnzlbg I don't know the details of the underlying architectures, unfortunately, but that is how loads and stores of the other types work. I'm not sure it's worth doing anything different for v128 without a concrete usecase in mind.

Reply:
We have to contend with C/C++ semantics here in addition to wasm semantics here, and in C/C++, it's UB if the pointer is unaligned in the current code.

I agree with @penzn that this operation should match the wasm instruction semantics. Clang's `<xmmintrin.h>` and in particular `__attribute__((__packed__, __may_alias__))` are some examples of ways to do unaligned-safe accesses with defined behavior.

Reply:
I think I'm slowly figuring this out.

I think what happens is that `memarg` passes an alignment hint, and the WASM SIMD spec states that for the loads and stores, that's the natural alignment of `v128`, which is 16. 

 The "front-end" (e.g. emscripten in this case) always has to pass 16 as the alignment hint for `v128.load/store`, even if it knows that the alignment could be less (e.g. if we were to add an `_unaligned` intrinsic here). 

This results in the WASM machine code generator having no idea that a load could be unaligned, and therefore being able to use an unaligned load instruction, which would be much faster than doing an aligned load, trapping, recovering, etc. 

cc @sunfishcode is this the case?

Reply:
It does not make sense that one cannot pass alignments smaller than the natural alignment in the hint. That has to actually be the only point about the hint in the first place. 

@tlively, the problem isn't that unaligned accesses are "slow". The problem is that unaligned accesses that have an incorrect alignment hint, e.g., one indicating that they are aligned when they are not, can be very slow. 

These intrinsics prevent the slowness by never letting users do an unaligned access. Rust (e.g. `packed_simd`) prevents the slowness by passing a correct alignment hint.

Reply:
> I think what happens is that `memarg` passes an alignment hint, and the WASM SIMD spec states that for the loads and stores, that's the natural alignment of `v128`, which is 16.

Yep!
 
> The "front-end" (e.g. emscripten in this case) always has to pass 16 as the alignment hint for `v128.load/store`, even if it knows that the alignment could be less (e.g. if we were to add an `_unaligned` intrinsic here).

Not quite. "Normal" loads and stores will get the natural alignment (16 in this case), but if you explicitly mark at the source level that some other alignment could be used then that will be reflected in the LLVM IR and respected in the alignment hints generated by the wasm backend.

> This results in the WASM machine code generator having no idea that a load could be unaligned, and therefore not being able to use an unaligned load instruction, which would be much faster than doing an aligned load, trapping, recovering, etc.

Since the alignment hints are only hints, engines have to be able to handle any load being potentially unaligned. But they might do this in the generated code if the alignment hint says a non-natural alignment will be used and do it only in traps (which would be faster in the common case) if the alignment hint says natural alignment will be used.

> Or is the alignment hint allowed to specify an alignment lower than the natural alignment?

Yes it is.

Reply:
+1 for having `wasm_v128_load()` and `wasm_v128_loadu()` for 16-byte aligned and 1-byte aligned (unaligned), or carrying the alignment hint 1-byte/4-byte/16-byte via an `__attribute__`. Unaligned loads and stores are very important from developers' point of view, there are times when one has to pack/unpack SIMD data to save memory, which equates to needing to do unaligned ops. One common use case is when doing fast loading/saving of files, where one does not want to carry any padding bytes around.

Reply:
@juj we currently have `wasm_v128_load` do unaligned loads, and if users want aligned loads they can just dereference a `v128_t*`. Do you think that is sufficient or would you like to see separate functions for aligned and unaligned accesses? If the latter, can you explain more why that would be better?

Reply:
That sounds good!